### ISSUE TITLE
[BUG] Inconsistent resistance with 'Get Gears from Zwift' setting in latest version of QZ.

### ISSUE THREAD

**Issue by Nagusia** [cagnulein/qdomyos-zwift#2680]
Hi Roberto.

The buttons on Zwift play behave as expected and the gear changes are displayed on the screen. However, with Zwift on ATV the resistance does not change as expected (there is a problem with the setup which might be the cause). With Zwift on Windows 11 the resistance does change but feels laggy and inconsistent.

Because of the bluetooth limitation of ATV I had to use a third device (Samsung Tab) with Zwift Companion in order to connect the Zwift Play controllers and the QZ app (on a Pixel 4A). In the end I was able to set things up as in your video and could see the gears changing but resistance did not change. I attach a debug log.

[debug-Sun_Oct_20_18_29_54_2024.log](https://github.com/user-attachments/files/17451714/debug-Sun_Oct_20_18_29_54_2024.log)

On Windows 11 it was easier to set up and I could feel the resistance changing although somewhat erratically. Occasionally it felt like the resistance just disappeared and then came back. It was not as consistent as the classic QZ behaviour, and not as easy to control. Is it possible in settings to adjust the baseline resistance, for gear 8 for example?

[debug-Sun_Oct_20_20_05_19_2024.log](https://github.com/user-attachments/files/17451707/debug-Sun_Oct_20_20_05_19_2024.log)

Is it possible to used the Zwift Click buttons with this new set up?

**Comment by cagnulein** [comment]
I  checked quickly the one on mobile. Set your trainer under ftms bike in
the bike settings and press OK and restart qz
Let me know!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 20 ott 2024 alle 20:59 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
@cagnulein Good morning. I just did a short ride with the change you suggested. All connects up okay and the gears change up and down on the screen with the Zwift Play buttons as expected, but there is no change to resistance I can detect. In fact, the resistance number on the QZ app remains more or less constant. Here is the debug file.

[debug-Mon_Oct_21_09_56_16_2024.log](https://github.com/user-attachments/files/17457454/debug-Mon_Oct_21_09_56_16_2024.log)

**Comment by cagnulein** [comment]
the problem in this log is that zwift is not synced with qz, zwift is
sending the hello message, qz is aswnering but zwift is still saying the
hello message again.
Can you show me a little video how you pair qz in zwift, i guess that your
issue is there.
Let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 21 ott 2024 alle ore 10:19 Nagusia ***@***.***>
ha scritto:

**Comment by Nagusia** [comment]
I have made a video of the connection process:
[QZ trial.zip](https://github.com/user-attachments/files/17461589/QZ.trial.zip)
It's not great quality I'm afraid but I hope you can see that I have connected everything correctly. You will also see that when I start the ride there is a 'connection failure' warning in Zwift. I go to the pairing screen to re-connect but although there is no longer a warning the resistance no longer changes. I guess that Zwift is not synchronised and imagine this is because I have to use Zwift companion to connect the phone running QZ as resistance source in Zwift. (This is because Zwift Play takes up the 2 available BT connections on ATV. I cannot see anywhere around this. Anyway, here are two debug files (I tried twice):
[debug-Mon_Oct_21_13_52_19_2024.log](https://github.com/user-attachments/files/17461735/debug-Mon_Oct_21_13_52_19_2024.log)
[debug-Mon_Oct_21_13_43_12_2024.log](https://github.com/user-attachments/files/17461818/debug-Mon_Oct_21_13_43_12_2024.log)
I did a further test running Zwift on a Samsung Tab which has no limitations on BT connections. This was easy to connect and gear changing appeared to work although, as with the Windows 11 experience yesterday, it was hard to feel the difference between 1 gear up or down and it felt laggy.  Here is the debug file:
[debug-Mon_Oct_21_14_52_39_2024.log](https://github.com/user-attachments/files/17461789/debug-Mon_Oct_21_14_52_39_2024.log)

**Comment by cagnulein** [comment]
> [debug-Mon_Oct_21_13_52_19_2024.log](https://github.com/user-attachments/files/17461735/debug-Mon_Oct_21_13_52_19_2024.log)

here i can see finally everything connected fine! but i don't see you changing gears

instead here https://github.com/user-attachments/files/17461818/debug-Mon_Oct_21_13_43_12_2024.log I can't see it, so maybe you did something different?





> changing appeared to work although, as with the Windows 11 experience yesterday, it was hard to feel the difference between 1 gear up or down and it felt laggy. Here is the debug file: [debug-Mon_Oct_21_14_52_39_2024.log](https://github.com/user-attachments/files/17461789/debug-Mon_Oct_21_14_52_39_2024.log)

the fact is laggy it's because you need also to enable the zwift gear experimental one in the zwift accessories in QZ. with this disabled, qz is adding 0.5% when you add a gear.

Follow this https://github.com/cagnulein/qdomyos-zwift/discussions/2671 i'm planning to change the zwift gear experimental in this way

**Comment by Nagusia** [comment]
> Follow this #2671 i'm planning to change the zwift gear experimental in this way

I've read the thread and the proposed table with 15 gears makes sense. I would be happy to help with testing on Android.

**Comment by Nagusia** [comment]
> the fact is laggy it's because you need also to enable the zwift gear experimental one in the zwift accessories in QZ. with this disabled, qz is adding 0.5% when you add a gear.

Despite enabling the Zwift gear experimental setting I am still not feeling consistent changes in resistance which make any sense. In a previous comment you said that my trainer - Wahoo Kickr v4 2018 - was not compatible with this setting. Is this still the case in this version 2.18.1?  Here is the debug file:
[debug-Mon_Oct_21_17_18_17_2024.log](https://github.com/user-attachments/files/17463432/debug-Mon_Oct_21_17_18_17_2024.log)

I also tried the ATV setup again but the constant disconnects with Zwift mean I can't really test anything.

**Comment by cagnulein** [comment]
> In a previous comment you said that my trainer - Wahoo Kickr v4 2018 - was not compatible with this setting.

hah you are right. Sorry I'm following too many cases. I will check again the log looking for an issue (probably tomorrow)

**Comment by cagnulein** [comment]
ok i can see an issue. Let me check deeply

**Comment by cagnulein** [comment]
no sorry, i was wrong, i see everything correctly.

just to understand better the issue:
- enable the target incline tile setting so you will see the target incline sent from zwift. Because I guess that when you say that it's sluggish it's becasue zwift is sending less inclination and so you feel like the gear isn't applied, but instead is the inclination that it's turning down

i checked the timing and from when you change gears from zwift to the timeslot when qz sends the frame to the trainer is just 100ms. The trainer then answers back after 400ms. So in total there is a 0.5s delay, but the most part is the trainer that it's not quick to act the command. Is this the delay that you are talking about?

Let me know your thoughts

**Comment by cagnulein** [comment]
also, last note: without the "get gears from zwift", do you have the same issue using the zwift play directly on qz?

**Comment by Nagusia** [comment]
> * enable the target incline tile setting so you will see the target incline sent from zwift. Because I guess that when you say that it's sluggish it's becasue zwift is sending less inclination and so you feel like the gear isn't applied, but instead is the inclination that it's turning down

Yes, I will do this. Now you mention it, this makes sense and confirms a suspicion I had that this is what might be happening. I guess the target incline tile will be an objective way on checking this. Should the target incline be the same as what Zwift shows?

> i checked the timing and from when you change gears from zwift to the timeslot when qz sends the frame to the trainer is just 100ms. The trainer then answers back after 400ms. So in total there is a 0.5s delay, but the most part is the trainer that it's not quick to act the command. Is this the delay that you are talking about?

No, I felt that lag was much longer. 0.5 s would be fine. I think I perceived the problem with inclination as lag.
 
> also, last note: without the "get gears from zwift", do you have the same issue using the zwift play directly on qz?

Not sure I understand this. I haven't used Zwift Play controllers directly on GZ because I didn't want to lose the other functions. I have been using Zwift Click with the previous version of GZ and the classic gear change system and it worked fine. (But it didn't work with the 'get gears from Zwift' setting - you said this was because my trainer wasn't compatible. Do you want me to try Zwift Click directly on GZ to see how it works?

**Comment by cagnulein** [comment]
> Yes, I will do this. Now you mention it, this makes sense and confirms a suspicion I had that this is what might be happening. I guess the target incline tile will be an objective way on checking this. Should the target incline be the same as what Zwift shows?

more or less. Zwift is showing the average one, qz is showing the actual one. Also there is a open point about this #2677 
so it could be this is the issue



> No, I felt that lag was much longer. 0.5 s would be fine. I think I perceived the problem with inclination as lag.

ok let me know



> . Do you want me to try Zwift Click directly on GZ to see how it works?

no it doesn't worth, you will see the same.



> I have been using Zwift Click with the previous version of GZ and the classic gear change system and it worked fine

perfect, this was the confirmation that i was looking, so the issue is about only to the "get gears from zwift" thing.

**Comment by Nagusia** [comment]
> more or less. Zwift is showing the average one, qz is showing the actual one. Also there is a open point about this #2677
> so it could be this is the issue

Actually, the new Zwift HUD shows the exact gradient at that moment in time, as well as the average. Anyway, I will let you know.

**Comment by cagnulein** [comment]
Didn't know this, thanks for the heads up!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 22 ott 2024 alle 15:35 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
I enabled the Target Incline Tile but it just showed 0% the whole time. I had the 'get gears from Zwift' setting enabled and also the 'use Zwift gears table' enabled (Is this right?). Starting off on the flat in gear 8 was comfortable but changing gear up or down didn't make any difference really. This continued to be the case on a 6% slope for a minute and then it suddenly got a lot harder despite changing to a lower gear. I turned round to go downhill and the resistance was maintained but then got easier again when I reached the flat. 
So there seem to be two problems: The Zwift gears table not working for my trainer. The incline issue.
[debug-Tue_Oct_22_15_58_31_2024.log](https://github.com/user-attachments/files/17477745/debug-Tue_Oct_22_15_58_31_2024.log)

**Comment by cagnulein** [comment]
> I had the 'get gears from Zwift' setting enabled and also the 'use Zwift gears table' enabled (Is this right?)

you can disable the "use zwift gears table", as I said it's not working for your traienr. But let's check the target incline thing, it shouldn't stay to 0

let me check the debug log

**Comment by cagnulein** [comment]
sorry my fault, it was the inclination tile, not the target one.
let me finish the log check anyway

**Comment by cagnulein** [comment]
@Nagusia i did a small patch https://github.com/cagnulein/qdomyos-zwift/actions/runs/11462610374
because when i was receiving the inclination from zwift, i wasn't adding any wind speed and any rolling and wind resistance. I was wondering if for your trainer these are required (this is happening only with get gears from zwift)

Google is very slow in these days, reviewing the app on the store, even the beta ones. For this reason, in order to check right now if the patch that I did is right, you can use the android build in the link (check at the bottom of the page in about 1 hour). In order to test it you need first to remove the current QZ version on your phone, install the one that you downloaded.
This version is a 15 minutes trial (you can restart it as many times as you want) and I can unlock it if you want, send me an email to roberto.viola83@gmail.com mentioning this ticket and that you are on android.

So to summarize:
- disable the use zwift gears table
- enable the incline tile and disable the target one
- try this new apk, if you didn't already send me an email to unlock the apk otherwise after 15 minutes it will close up

Let me know!

**Comment by Nagusia** [comment]
Will do. Just one question. Can I save the profile in present app before uninstalling and then import in the new version?  When I tried this with the most recent version 2.18.1 I seemed to lose all the settings.

**Comment by cagnulein** [comment]
you can but follow this:
- save the profile
- move the profile file somewhere else outside qz
- uninstall the version
- install the apk
- remove the qdomyos folder
- run qz and import the profile directly from the profile windows in qz, pressing the other folder button

let me know!

**Comment by Nagusia** [comment]
While waiting for the new build link did a ride with everything connected to virtual wahoo trainer (including resistance) and with the inclination tile enabled. The figure in the tile was identical to Zwift (climb difficulty at max) but half the Zwift percentage on descents (this is built in to Zwift I believe). 

> * disable the use zwift gears table
> * enable the incline tile and disable the target one

I just want to confirm that with the new build you want me to trial it with 'get gears from Zwift' enabled (as opposed to 'use Zwift gears') and resistance connected directly to phone with QZ on, and not the virtual trainer.

**Comment by cagnulein** [comment]
> While waiting for the new build link did a ride with everything connected to virtual wahoo trainer (including resistance) and with the inclination tile enabled. The figure in the tile was identical to Zwift (climb difficulty at max) but half the Zwift percentage on descents (this is built in to Zwift I believe).

yes for the negative inclination is zwift that send the half. there is a setting on qz to double this.

this is expected because zwift, for the devices compatible with the cog, is not sending the inclination but its sending a sort of effort. i added an algorithm to estimate the inclination, its why we have to understand what i could improve 


> 
> 
> 
> > * disable the use zwift gears table
> 
> > * enable the incline tile and disable the target one
> 
> 
> 
> I just want to confirm that with the new build you want me to trial it with 'get gears from Zwift' enabled (as opposed to 'use Zwift gears') and resistance connected directly to phone with QZ on, and not the virtual trainer.
> 
> 
> 
> 

exactly

**Comment by Nagusia** [comment]
Tested the new build. The incline tile figure doesn't match up with Zwift. On uphill it is about half the Zwift figure. On downhills I didn't see a negative figure. 
You start off on gear 8 in Zwift, which corresponds to gear 0 in QZ. If you change to gear 7 in Zwift, it jumps to -8 in QZ. If you use the Zwift Click this doesn't happen and I guess it changes the incline 0.5% with each gear change as expected. 
[debug-Tue_Oct_22_18_38_47_2024.log](https://github.com/user-attachments/files/17479726/debug-Tue_Oct_22_18_38_47_2024.log)

**Comment by cagnulein** [comment]
yes but beside the fact that it's not following exactly the inclination and that the gears does a jump on startup, now do you feel the resistance following the gears? the modification that i did in this one it's about this

**Comment by Nagusia** [comment]
> now do you feel the resistance following the gears?

Yes, I think so. It's certainly better than before, although I'm afraid I was more focused on the inclination issue. I'll try to give you more feedback tomorrow.

**Comment by cagnulein** [comment]
Ok I will work on the inclination side

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 22 ott 2024 alle 19:25 Nagusia ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
i fixed the inclination! on both negative and positive with the get gears,
i will upload a new one soon
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 22 ott 2024 alle ore 20:03 cagnulein ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
try this https://github.com/cagnulein/qdomyos-zwift/actions/runs/11475844697
now the inclination is perfect!

**Comment by Nagusia** [comment]
The inclination is improved but uphill still marks about half the inclination figure in Zwift. Downhill it shows negative inclination but again, doesn't match Zwift. On a -6% slope it showed -1-1.5% maybe. It doesn't work as well as the setup where 'get gears from Zwift' is disabled. It may be that I haven't understood what you're aiming for here.
As far as the gears are concerned, there is definitely a consistent change in resistance, but it works better with Zwift Click because there are no jumps. As with the Zwift Play controller you can't go down from 0 (it jumps to -8) it not possible to feel the gears/resistance change in the way you can with Zwift Click.
At present, if I was going to do a serious training ride, I would use the non 'get gears from Zwift' setup, and use Zwift Click.
[debug-Wed_Oct_23_16_08_21_2024.log](https://github.com/user-attachments/files/17493657/debug-Wed_Oct_23_16_08_21_2024.log)
I have a question: I have been looking at the #2671 thread and see you are making progress there. Is the idea of using wheel circumference in a table to apply gears going to be possible for my trainer - Wahoo Kickr v4? (So far it's not possible to use Zwift tables because it is not compatible.) If it is, I would be happy to help with testing.

**Comment by cagnulein** [comment]
First of all did you set the difficulties at 100% on zwift?  What you are
describing uphill it's about this
Let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 23 ott 2024 alle 16:52 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
Sorry. Yes, I was using Zwift on my tab and the climb difficulty setting was at 50 percent. Just done another test and inclination is good now - matching on uphill and 50% on downhill. I think each gear change needs to be 1% rather than 0.5% or you are going to run out of options on steep climbs of 13%.
[debug-Wed_Oct_23_18_31_53_2024.log](https://github.com/user-attachments/files/17495653/debug-Wed_Oct_23_18_31_53_2024.log)

**Comment by cagnulein** [comment]
Simply set the gears gain to 2 to obtain this!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 23 ott 2024 alle 18:51 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
> Simply set the gears gain to 2 to obtain this!

Yes, but how will that work out with Zwift gears (assuming the jump problem can be solved)?  Will changing one gear in Zwift change 2 in GZ?

**Comment by cagnulein** [comment]
Yes exactly!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 23 ott 2024 alle 19:10 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
OK. And the jump problem?

**Comment by cagnulein** [comment]
Which is the jump issue?

**Comment by Nagusia** [comment]
When I use the Zwift Play controller to change from 8 to 7, GZ changes from 0 to -8.

**Comment by cagnulein** [comment]
Hah right the first time? I will fix this tomorrow

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 23 ott 2024 alle 19:23 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
Great. Thanks Roberto.

**Comment by cagnulein** [comment]
Fixed
https://github.com/cagnulein/qdomyos-zwift/actions/runs/11486312759


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 23 ott 2024 alle 19:38 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
I'm afraid not. Now 8 to 7 on Zwift gears is 0 to -16 in QZ gears tile. Works alright with Zwift Click though.

**Comment by cagnulein** [comment]
Please show me a fresh debug log, maybe i did a mistake (for sure)
Thanks

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 24 ott 2024 alle 13:51 Nagusia ***@***.***> ha
scritto:

**Comment by Nagusia** [comment]
-16 instead of -8 because I've set gear gain to 2 I guess.

**Comment by cagnulein** [comment]
ok show me a debug log of the issue and I will check it

**Comment by cagnulein** [comment]
ok there was 2 errors

https://github.com/cagnulein/qdomyos-zwift/actions/runs/11500132647

now it should be fixed!

**Comment by cagnulein** [comment]
sorry new link https://github.com/cagnulein/qdomyos-zwift/actions/runs/11500673128

**Comment by Nagusia** [comment]
Yes. It works perfectly now on Android. Just wish I could find a way to make it work with ATV. The resistance control doesn't work well connected via Zwift Companion. Is there any way to make the resistance control DC (WiFi) like power and cadence?

**Comment by cagnulein** [comment]
> Is there any way to make the resistance control DC (WiFi) like power and cadence?

i would need a log from a ethernet traienr with full zwift play support for this



> The resistance control doesn't work well connected via Zwift Companion

i have a user that achieved this. it's on messenger. if you want i can put you guys in contact

**Comment by cagnulein** [comment]
BTW can i close this so?

**Comment by Nagusia** [comment]
> i have a user that achieved this. it's on messenger. if you want i can put you guys in contact

Yes please.

> BTW can i close this so?

Yes, I guess so. Thanks very much.

**Comment by cagnulein** [comment]
I remember wrong, he came to your same conclusions. So we have to find someone with a ethernet device and play compatible. Are you aware if there is any device with it?

**Comment by cagnulein** [comment]
Maybe the new jet victory (if I got the right name)

**Comment by cagnulein** [comment]
yes this one https://www.jetblackcycling.com/indoor-cycle-trainers-and-accessories/victory-smart-trainer/

**Comment by Nagusia** [comment]
Maybe I should ask on Reddit if there is someone with this model who could help.

**Comment by cagnulein** [comment]
> Maybe I should ask on Reddit if there is someone with this model who could help.

already did :)
